File: Packages/TypeEngine/Sources/TypeEngine/ProxySimulator.swift
import Foundation

/// Headless model of the iOS `UITextDocumentProxy` contract, for the macOS
/// harness (`type-repl`) and unit tests. Real keyboard extensions never see
/// the whole document: they get a truncated window (`documentContextBefore/
/// AfterInput`), can only `insertText`/`deleteBackward` at the cursor, and
/// must survive cursor jumps, host-app text mutation, and briefly-stale
/// context reads after inserts. This type reproduces those constraints so
/// `TypingSession` can be exercised against them off-device.
///
/// Modeled behaviors:
/// - truncated `contextBeforeInput` (configurable policy; default mimics the
///   common iOS shape: cut at the most recent sentence terminator — ./!/?
///   followed by a space — or a newline, and cap at 200 characters,
///   whichever window is shorter)
/// - `insertText` / `deleteBackward` at the cursor only
/// - cursor moves and wholesale host-app text replacement
/// - optional stale reads: the first context read after an `insertText`
///   returns the pre-insert state (real proxies are briefly stale)
///
/// Documented out of scope (device-tested, host-app-specific): per-app
/// window sizes and boundary quirks (hosts differ wildly), multi-stage
/// asynchronous context refresh, marked text / IME composition, and
/// selection ranges (only a caret is modeled).
public final class ProxySimulator {

    // MARK: - Truncation policy

    /// How `contextBeforeInput` is cut down from the full text before the
    /// cursor. Real iOS behavior varies by host app; the default here is the
    /// common shape (current sentence, capped length). Use `custom` to model
    /// a specific host.
    public struct TruncationPolicy {
        /// Hard cap on the returned window, in characters (applied last).
        public var maxBeforeLength: Int
        /// Cut at the most recent sentence terminator (". ", "! ", "? ") or
        /// newline, returning only the text after it.
        public var cutAtSentenceBoundary: Bool
        /// Full override: given the complete text before the cursor, return
        /// the window the proxy exposes. When set, the other fields are
        /// ignored.
        public var custom: ((String) -> String)?

        public init(
            maxBeforeLength: Int = 200,
            cutAtSentenceBoundary: Bool = true,
            custom: ((String) -> String)? = nil
        ) {
            self.maxBeforeLength = maxBeforeLength
            self.cutAtSentenceBoundary = cutAtSentenceBoundary
            self.custom = custom
        }

        /// The whole text before the cursor, no truncation (for tests).
        public static let none = TruncationPolicy(
            maxBeforeLength: .max,
            cutAtSentenceBoundary: false
        )

        func apply(to full: String) -> String {
            if let custom { return custom(full) }
            var window = Substring(full)
            if cutAtSentenceBoundary {
                window = Self.currentSentence(of: window)
            }
            if window.count > maxBeforeLength {
                window = window.suffix(maxBeforeLength)
            }
            return String(window)
        }

        /// Text after the most recent sentence terminator (./!/? followed by
        /// a space) or newline; the whole text when there is none.
        private static func currentSentence(of text: Substring) -> Substring {
            var boundary: Substring.Index?
            var previous: Character?
            var index = text.startIndex
            while index < text.endIndex {
                let ch = text[index]
                if ch.isNewline {
                    boundary = text.index(after: index)
                } else if ch == " ", let previous, ".!?".contains(previous) {
                    boundary = text.index(after: index)
                }
                previous = ch
                index = text.index(after: index)
            }
            guard let boundary else { return text }
            return text[boundary...]
        }
    }

    // MARK: - State

    /// The full document (what the host app holds; the keyboard never sees
    /// all of it).
    public private(set) var document: String
    /// Caret position as a character offset into `document` (0...count).
    public private(set) var cursor: Int

    public var truncation: TruncationPolicy

    /// When true, the first context read after each `insertText` /
    /// `deleteBackward` returns the pre-edit state — modeling the brief
    /// staleness of real proxies. Off by default.
    public var staleReads: Bool = false
    /// Pending stale snapshot: (before, after) windows captured pre-edit.
    private var staleSnapshot: (before: String, after: String)?

    public init(
        document: String = "",
        cursorAt cursor: Int? = nil,
        truncation: TruncationPolicy = TruncationPolicy()
    ) {
        self.document = document
        self.cursor = min(cursor ?? document.count, document.count)
        self.truncation = truncation
    }

    // MARK: - Context windows (what the keyboard sees)

    /// Truncated text before the cursor — the analogue of
    /// `documentContextBeforeInput`. Consumes the stale snapshot if one is
    /// pending.
    public var contextBeforeInput: String {
        if let stale = takeStaleSnapshotIfPending() { return stale.before }
        return truncation.apply(to: fullTextBeforeCursor)
    }

    /// Text after the cursor — the analogue of `documentContextAfterInput`
    /// (unbounded here; after-window truncation is not load-bearing for the
    /// engine, which only consumes the before-window).
    public var contextAfterInput: String {
        if let stale = takeStaleSnapshotIfPending() { return stale.after }
        return fullTextAfterCursor
    }

    /// Read both windows in one consistent snapshot (a single stale snapshot
    /// covers both, like one proxy read).
    public func contextWindows() -> (before: String, after: String) {
        if let stale = takeStaleSnapshotIfPending() { return stale }
        return (truncation.apply(to: fullTextBeforeCursor), fullTextAfterCursor)
    }

    // MARK: - Edits (all the keyboard can do)

    public func insertText(_ text: String) {
        captureStaleSnapshotIfEnabled()
        let index = characterIndex(at: cursor)
        document.insert(contentsOf: text, at: index)
        cursor += text.count
    }

    public func deleteBackward() {
        guard cursor > 0 else { return }
        captureStaleSnapshotIfEnabled()
        let index = characterIndex(at: cursor - 1)
        document.remove(at: index)
        cursor -= 1
    }

    // MARK: - Things that happen TO the keyboard

    /// Move the caret to an absolute character offset (cursor jump: user
    /// tapped elsewhere in the text). Clamped to the document bounds.
    public func moveCursor(to offset: Int) {
        cursor = min(max(offset, 0), document.count)
        staleSnapshot = nil
    }

    /// Move the caret by a relative delta.
    public func moveCursor(by delta: Int) {
        moveCursor(to: cursor + delta)
    }

    /// The host app replaces the text under us (autofill, undo, programmatic
    /// set). Cursor goes to `cursorAt` (default: end of new text).
    public func hostReplaceText(_ newDocument: String, cursorAt: Int? = nil) {
        document = newDocument
        cursor = min(cursorAt ?? newDocument.count, newDocument.count)
        staleSnapshot = nil
    }

    // MARK: - Internals

    private var fullTextBeforeCursor: String {
        String(document.prefix(cursor))
    }

    private var fullTextAfterCursor: String {
        String(document.suffix(document.count - cursor))
    }

    private func characterIndex(at offset: Int) -> String.Index {
        document.index(document.startIndex, offsetBy: offset)
    }

    private func captureStaleSnapshotIfEnabled() {
        guard staleReads else { return }
        staleSnapshot = (truncation.apply(to: fullTextBeforeCursor), fullTextAfterCursor)
    }

    private func takeStaleSnapshotIfPending() -> (before: String, after: String)? {
        guard let snapshot = staleSnapshot else { return nil }
        staleSnapshot = nil
        return snapshot
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/ProxySimulator.kt
package `is`.solberg.lyklabord.engine

/**
 * Headless model of the iOS UITextDocumentProxy contract used by the parity
 * harness. Context before the cursor is truncated, while edits are limited to
 * insertText/deleteBackward at the caret.
 */
class ProxySimulator(
    document: String = "",
    cursorAt: Int? = null,
    truncation: TruncationPolicy = TruncationPolicy(),
) {

    /** Policy used to truncate the text before the cursor. */
    class TruncationPolicy(
        var maxBeforeLength: Int = 200,
        var cutAtSentenceBoundary: Boolean = true,
        var custom: ((String) -> String)? = null,
    ) {
        /** The whole text before the cursor, without truncation. */
        companion object {
            val none: TruncationPolicy
                get() = TruncationPolicy(
                    maxBeforeLength = Int.MAX_VALUE,
                    cutAtSentenceBoundary = false,
                )
        }

        internal fun apply(to: String): String {
            val override = custom
            if (override != null) return override(to)
            var window = to
            if (cutAtSentenceBoundary) {
                window = currentSentence(window)
            }
            return if (window.length > maxBeforeLength) {
                window.takeLast(maxBeforeLength)
            } else {
                window
            }
        }

        /** Text after the most recent terminator or newline. */
        private fun currentSentence(text: String): String {
            var boundary: Int? = null
            var previous: Char? = null
            for (index in text.indices) {
                val ch = text[index]
                if (ch.isNewline()) {
                    boundary = index + 1
                } else if (ch == ' ' && previous != null && previous in ".!?") {
                    boundary = index + 1
                }
                previous = ch
            }
            return boundary?.let { text.substring(it) } ?: text
        }

        private fun Char.isNewline(): Boolean = when (this) {
            '\n', '\r', '\u0085', '\u2028', '\u2029' -> true
            else -> false
        }
    }

    /** A named pair mirroring Swift's `(before: String, after: String)` tuple. */
    data class ContextWindows(val before: String, val after: String)

    /** The full host document. */
    var document: String = document
        private set
    /** Caret offset into [document]. */
    var cursor: Int = minOf(cursorAt ?: document.length, document.length)
        private set

    var truncation: TruncationPolicy = truncation

    /** Return one pre-edit observation after each keyboard edit when enabled. */
    var staleReads: Boolean = false
    private var staleSnapshot: ContextWindows? = null

    /** Restore keyboard edits on the next observation read when enabled. */
    var swallowEdits: Boolean = false
    private var swallowRestore: SwallowRestore? = null

    /** Snapshot captured before a swallowed edit batch. */
    private data class SwallowRestore(val document: String, val cursor: Int)

    /** Truncated analogue of UITextDocumentProxy.documentContextBeforeInput. */
    val contextBeforeInput: String
        get() {
            applySwallowRestoreIfPending()
            takeStaleSnapshotIfPending()?.let { return it.before }
            return truncation.apply(to = fullTextBeforeCursor)
        }

    /** Analogue of UITextDocumentProxy.documentContextAfterInput. */
    val contextAfterInput: String
        get() {
            applySwallowRestoreIfPending()
            takeStaleSnapshotIfPending()?.let { return it.after }
            return fullTextAfterCursor
        }

    /** Read both context windows from one consistent snapshot. */
    fun contextWindows(): ContextWindows {
        applySwallowRestoreIfPending()
        return takeStaleSnapshotIfPending()
            ?: ContextWindows(truncation.apply(to = fullTextBeforeCursor), fullTextAfterCursor)
    }

    /** Ground-truth truncated before-window after the keyboard's own edits. */
    val trueContextBeforeInput: String
        get() = truncation.apply(to = fullTextBeforeCursor)

    fun insertText(text: String) {
        captureStaleSnapshotIfEnabled()
        captureSwallowRestoreIfEnabled()
        val index = characterIndex(at = cursor)
        document = document.substring(0, index) + text + document.substring(index)
        cursor += text.length
    }

    fun deleteBackward() {
        if (cursor <= 0) return
        captureStaleSnapshotIfEnabled()
        captureSwallowRestoreIfEnabled()
        val index = characterIndex(at = cursor - 1)
        document = document.removeRange(index, index + 1)
        cursor -= 1
    }

    /** Move the caret to an absolute offset or by a relative delta. */
    fun moveCursor(to: Int? = null, by: Int? = null) {
        require((to == null) xor (by == null)) { "exactly one cursor movement must be specified" }
        val offset = to ?: (cursor + by!!)
        cursor = minOf(maxOf(offset, 0), document.length)
        staleSnapshot = null
        swallowRestore = null
    }

    /** Replace the host document and optionally choose the new caret offset. */
    fun hostReplaceText(newDocument: String, cursorAt: Int? = null) {
        document = newDocument
        cursor = minOf(cursorAt ?: newDocument.length, newDocument.length)
        staleSnapshot = null
        swallowRestore = null
    }

    private val fullTextBeforeCursor: String
        get() = document.substring(0, cursor)

    private val fullTextAfterCursor: String
        get() = document.substring(cursor)

    private fun characterIndex(at: Int): Int = at

    private fun captureStaleSnapshotIfEnabled() {
        if (!staleReads) return
        staleSnapshot = ContextWindows(
            before = truncation.apply(to = fullTextBeforeCursor),
            after = fullTextAfterCursor,
        )
    }

    private fun captureSwallowRestoreIfEnabled() {
        if (!swallowEdits || swallowRestore != null) return
        swallowRestore = SwallowRestore(document, cursor)
    }

    private fun applySwallowRestoreIfPending() {
        val restore = swallowRestore ?: return
        swallowRestore = null
        document = restore.document
        cursor = restore.cursor
    }

    private fun takeStaleSnapshotIfPending(): ContextWindows? {
        val snapshot = staleSnapshot ?: return null
        staleSnapshot = null
        return snapshot
    }
}
